core/src/repair/quic_endpoint.rs

```diff
@@ -0,0 +1,624 @@
+    cache: Arc<RwLock<ConnectionCache>>,
+) {
+    if let Err(err) = handle_connecting(endpoint, connecting, remote_request_sender, cache).await {
+        error!("handle_connecting: {err:?}");
+    }
+}
+
+async fn handle_connecting(
+    endpoint: Endpoint,
+    connecting: Connecting,
+    remote_request_sender: Sender<RemoteRequest>,
+    cache: Arc<RwLock<ConnectionCache>>,
+) -> Result<(), Error> {
+    let connection = connecting.await?;
+    let remote_address = connection.remote_address();
+    let remote_pubkey = get_remote_pubkey(&connection)?;
+    handle_connection_error(
+        endpoint,
+        remote_address,
+        remote_pubkey,
+        connection,
+        remote_request_sender,
+        cache,
+    )
+    .await;
+    Ok(())
+}
+
+async fn handle_connection_error(
+    endpoint: Endpoint,
+    remote_address: SocketAddr,
+    remote_pubkey: Pubkey,
+    connection: Connection,
+    remote_request_sender: Sender<RemoteRequest>,
+    cache: Arc<RwLock<ConnectionCache>>,
+) {
+    cache_connection(remote_address, remote_pubkey, connection.clone(), &cache).await;
+    if let Err(err) = handle_connection(
+        &endpoint,
+        remote_address,
+        remote_pubkey,
+        &connection,
+        &remote_request_sender,
+    )
+    .await
+    {
+        drop_connection(remote_address, remote_pubkey, &connection, &cache).await;
+        error!("handle_connection: {remote_pubkey}, {remote_address}, {err:?}");
+    }
+}
+
+async fn handle_connection(
+    endpoint: &Endpoint,
+    remote_address: SocketAddr,
+    remote_pubkey: Pubkey,
+    connection: &Connection,
+    remote_request_sender: &Sender<RemoteRequest>,
+) -> Result<(), Error> {
+    loop {
+        let (send_stream, recv_stream) = connection.accept_bi().await?;
+        tokio::task::spawn(handle_streams_task(
+            endpoint.clone(),
+            remote_address,
+            remote_pubkey,
+            send_stream,
+            recv_stream,
+            remote_request_sender.clone(),
+        ));
+    }
+}
+
+async fn handle_streams_task(
+    endpoint: Endpoint,
+    remote_address: SocketAddr,
+    remote_pubkey: Pubkey,
+    send_stream: SendStream,
+    recv_stream: RecvStream,
+    remote_request_sender: Sender<RemoteRequest>,
+) {
+    if let Err(err) = handle_streams(
+        &endpoint,
+        remote_address,
+        remote_pubkey,
+        send_stream,
+        recv_stream,
+        &remote_request_sender,
+    )
+    .await
+    {
+        error!("handle_stream: {remote_address}, {remote_pubkey}, {err:?}");
+    }
+}
+
+async fn handle_streams(
+    endpoint: &Endpoint,
+    remote_address: SocketAddr,
+    remote_pubkey: Pubkey,
+    mut send_stream: SendStream,
+    mut recv_stream: RecvStream,
+    remote_request_sender: &Sender<RemoteRequest>,
+) -> Result<(), Error> {
+    // Assert that send won't block.
+    debug_assert_eq!(remote_request_sender.capacity(), None);
+    let bytes = recv_stream.read_to_end(PACKET_DATA_SIZE).await?;
```

Comment:
What if the client do not give all the data intentionally? Do we need timeout?

Reply:
added timeout.